### PR TITLE
[PD] Optimize paged allocator free-list release

### PR DESCRIPTION
**Pull request by wangwenmingaa**
## Motivation

This PR is a follow-up to #33998, which defers free-list concatenation for
`LogicalHostPool`.

`PagedTokenToKVPoolAllocator` still eagerly concatenates released page IDs on
every `free()` when `need_sort=True`:

```python
self.release_pages = torch.cat((*page_ids, self.release_pages))
```

This path is used by PD disaggregation allocators. Under workloads with frequent
SWA page releases, every release reallocates and copies the growing
`release_pages` tensor, even though these pages are not reused until the primary
free list runs low.

If the current release list contains `F` pages and the current call releases
`K` pages, each release performs `O(F + K)` free-list maintenance. Repeating
this operation during cache eviction introduces unnecessary allocator overhead.

This PR moves the concatenation out of the release hot path. Released page
tensors are queued as chunks and materialized only when an allocation cannot be
satisfied by the primary free list.

## Modifications

Apply deferred free-list merging to `PagedTokenToKVPoolAllocator` when
`need_sort=True`:

- Keep released page-ID tensors in `pending_release_page_chunks` instead of
  concatenating them immediately.
- Track the number of pending released pages with
  `num_pending_release_pages`.
- Include pending pages in `available_size()`.
- Continue allocating from the primary `free_pages` tensor while it has
  sufficient capacity.
- Concatenate and sort pending chunks only when allocation pressure requires
  the released pages.
- Avoid recording empty releases.
- Reset all pending-release state in `clear()`.
- Include pending pages in duplicate-page and memory-leak invariant checks.
- Preserve the existing sorted page-reuse behavior.

The behavior for allocators with `need_sort=False` is unchanged.

The patch does not add runtime logging, configuration flags, or model-specific
behavior.

Unit tests cover:

- deferred release without rebuilding the primary free list;
- allocation from the primary free list before merging pending pages;
- sorted reuse of released pages after a deferred merge;
- merging multiple released chunks;
- pending-page availability accounting;
- state reset through `clear()`;
- preservation of the existing non-PD allocation behavior;
- duplicate-page detection across pending releases.

## Accuracy Tests

This change only modifies allocator free-list metadata.

It does not modify:

- model weights or model forward execution;
- KV cache tensor contents;
- cache keys or page-table mappings;
- attention, sampling, or CUDA kernels;
- PD KV transfer contents or protocols.

The allocator regression tests verify that the same page IDs are released and
reused in sorted order. Therefore, model output accuracy is not expected to
change.

Added and updated tests:

```text
test/registered/unit/mem_cache/test_paged_allocator_lazy_release.py
test/registered/unit/mem_cache/test_paged_free_segment.py
```

Local checks completed:

```text
PASS: Python syntax compilation for the modified source and test files
PASS: scripts/lint/check_registered_tests.py
PASS: git diff --check
```

The registered CPU unit tests will also be executed by GitHub CI:

```bash
python3 test/registered/unit/mem_cache/test_paged_allocator_lazy_release.py
python3 test/registered/unit/mem_cache/test_paged_free_segment.py
```

## Speed Tests and Profiling

Before this change, every release may concatenate new page IDs with a growing
release list:

```text
Release-side free-list maintenance: O(F + K)
```

After this change, `free()` only records the released tensors and updates a
counter:

```text
Release-side free-list maintenance: O(1)
```

The larger concatenation and sort are deferred until the primary free list
cannot satisfy an allocation. Multiple releases can therefore be combined into
one allocation-side merge.

| Behavior | Before | After |
| --- | --- | --- |
| `torch.cat` during every release | Yes | No |
| Primary allocation fast path | Tensor slice | Tensor slice |
| Released-page accounting | Eager tensor rebuild | Append chunk and update counter |
| Large merge | Once per release | Only when released pages are needed |
| Page reuse order | Sorted before reuse | Unchanged |
| `need_sort=False` behavior | Existing behavior | Unchanged |

The allocator was profiled on a DeepSeek V4 Flash PD prefill workload with
CP=4 and page size 256. Baseline and optimized runs used the same request
workload and serving configuration. The following results aggregate comparable
steady-state windows across four CP ranks.

### SWA allocator

| Metric | Eager baseline | Deferred merge | Change |
| --- | ---: | ---: | ---: |
| Release calls | 5,792 | 6,004 | Comparable load |
| Released pages | 229,652 | 238,720 | Comparable load |
| Release-time `torch.cat` calls | 5,792 | 0 | -100% |
| CPU release-list update | 17.544 us/call | 1.506 us/call | -91.4% |
| CPU amortized queue cost | 20.590 us/call | 4.987 us/call | -75.8% |
| GPU release-list update | 0.03076 ms/call | 0.01202 ms/call | -60.9% |
| GPU amortized queue cost | 0.03514 ms/call | 0.01698 ms/call | -51.7% |

### Full-attention allocator

| Metric | Eager baseline | Deferred merge | Change |
| --- | ---: | ---: | ---: |
| Release calls | 5,968 | 6,016 | Comparable load |
| Released pages | 237,208 | 239,224 | Comparable load |
| Release-time `torch.cat` calls | 5,968 | 0 | -100% |
| CPU release-list update | 19.041 us/call | 1.517 us/call | -92.0% |
| CPU amortized queue cost | 41.373 us/call | 32.900 us/call | -20.5% |
| GPU release-list update | 0.02866 ms/call | 0.01175 ms/call | -59.0% |
| GPU amortized queue cost | 0.06039 ms/call | 0.05236 ms/call | -13.3% |

The SWA allocator benefits more because multiple out-of-window releases can
accumulate before the released pages are needed. Full-attention pages are more
often evicted immediately before an allocation retry, so pending chunks are
merged more frequently.

These measurements isolate allocator hot-path overhead. This PR does not claim
an end-to-end throughput improvement.

## Checklist

- [x] Add unit tests according to the
  [unit-test guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Documentation is not required because this change does not introduce or
  modify a user-facing interface.
- [x] Provide accuracy and performance analysis.
- [x] Follow the SGLang code-style guidance.

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the
   [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from
   [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS)
   and other reviewers.
3. Trigger CI tests with the documented CI comments or contact an authorized
   user.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, and
     `/rerun-failed-ci`.
4. After CI is green and required approvals are collected, ask Merge Oncalls or
   a maintainer with write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #33694881968](https://github.com/sgl-project/sglang/actions/runs/33694881968)
Latest PR Test (Extra): :x: [Run #33694881807](https://github.com/sgl-project/sglang/actions/runs/33694881807)
Latest PR Test (AMD ROCm 7.2): :hourglass_flowing_sand: [Run #33694881930](https://github.com/sgl-project/sglang/actions/runs/33694881930)

### CODE REVIEW

**Comment by wangwenmingaa** [comment]
@hzh0425  Could you please review this PR when you have a chance?

This PR optimizes free-list release in `PagedTokenToKVPoolAllocator` by deferring repeated `torch.cat` operations and merging released indices in batches. It follows the same general idea as #33998, but applies it to the device-side paged allocator.

Profiling shows that the change removes `torch.cat` from the release path and reduces the amortized SWA allocator queue-management cost by 75.8% on CPU and 51.7% on GPU, while preserving allocation order and allocator behavior.

Since you are a code owner for the allocator module, your feedback would be greatly appreciated. Thank you!

**Comment by wangwenmingaa** [comment]
@hnyls2002 Hi, could you please take a look at this PR when you have a chance?

It is still waiting for a review from someone familiar with the allocator path. The PR batches released indices in `PagedTokenToKVPoolAllocator`, removing repeated `torch.cat` operations from the release path. Profiling shows a 75.8% CPU and 51.7% GPU reduction in the amortized SWA allocator queue-management cost, while preserving the existing allocation order and behavior.

I previously requested a review, but the PR was assigned to a different reviewer, so I’m reaching out to you as well. Your feedback would be greatly appreciated. Thank you!

**Review by hnyls2002 — APPROVED** [review]

**Comment by hnyls2002** [comment]
/rerun-test test_paged_allocator_lazy_release.py test_paged_free_segment.py test_kv_page_invariants.py test_minicpm_sparse_cache.py test_dcp_layout_unit.py test/registered/disaggregation/test_disaggregation_basic.py

**Comment by hnyls2002** [comment]
/rerun-test registered/disaggregation/test_disaggregation_basic.py
